Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -681,37 +681,55 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
                     if (!light.is_enabled)
                         continue;
 
-                    FloatVector4 vertex_to_light;
+                    // We need to save the length here because the attenuation factor requires a non
+                    // normalized vector!
+                    auto sgi_arrow_operator = [](FloatVector4 const& p1, FloatVector4 const& p2, float& saved_length) {
+                        if ((p1.w() != 0.0f) && (p2.w() == 0.0f)) {
+                            saved_length = p2.length();
+                            return (p2 / saved_length).xyz();
+                        } else if ((p1.w() == 0.0f) && (p2.w() != 0.0f)) {
+                            saved_length = p2.length();
+                            return -((p1 / saved_length)).xyz();
```

Comment:
Are these extra parentheses necessary?

Reply:
Probably not, but it makes it more explicit what operation is being performed first.